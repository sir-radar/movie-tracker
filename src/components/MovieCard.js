import React from 'react';

const MovieCard = ({id, title, image, overview,addToFavorite, removeFavorite, favoriteIDs}) => {

  const truncate = (str) => str.length > 100 ? str.substring(0, 100) + "..." : str;

  const makeFavourite = (e) =>{
    const data = {
      "media_type": "movie",
      "media_id": id,
      "favorite": true
    };
    e.preventDefault()
    addToFavorite(data)
  }

  const removeFavourite = () => {
    const data = {
      "media_type": "movie",
      "media_id": id,
      "favorite": false
    };
    removeFavorite(data)
  }

  return (
    <div className="card card-height p-0 col-12 col-md-5 m-2">
      <img src={image ? `https://image.tmdb.org/t/p/w200/${image}` : 'no-image.png'} className="card-img-top image-size" alt="..."/>
      <div className="d-flex flex-column justify-content-end card-body">
        <h5 className="card-title">{title}</h5>
        <div>
          <p className="card-text">{truncate(overview)}</p>
        </div>
        <div className="align-self-center actions">
          <a href="#" className="btn btn-primary">Go somewhere</a>
          <div className="d-flex">
            {
              favoriteIDs.includes(id) ? <button onClick={removeFavourite} className="btn mr-2"><i className="fa fa-star text-favourite"></i></button>
                : <button onClick={makeFavourite} className="btn mr-2"><i className="fa fa-star"></i></button>
            }
            
            <button className="btn">Watch later</button>
          </div>
        </div>
      </div>
    </div>
  );
}

export default MovieCard;
