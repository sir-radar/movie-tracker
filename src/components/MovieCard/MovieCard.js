import React from 'react';
import {Link} from 'react-router-dom';

const MovieCard = (props) => {

  const { id,
          title, 
          image, 
          overview, 
          addToFavourite, 
          removeFavourite, 
          favouriteIDs,
          removeFromWatchList,
          addToWatchList,
          watchlistIDs
        } = props

  //Limits text to the length
  const truncate = (str, length) => str.length > length ? str.substring(0, length) + "..." : str;

  const makeFavouriteHandler = (e) =>{
    const data = {
      "media_type": "movie",
      "media_id": id,
      "favorite": true
    };
    e.preventDefault()
    addToFavourite(data)
  }

  const removeFromFavouritesHandler = () => {
    const data = {
      "media_type": "movie",
      "media_id": id,
      "favorite": false
    };
    removeFavourite(data)
  }

  const addToWatchListHandler = () => {
    const data = {
      "media_type": "movie",
      "media_id": id,
      "watchlist": true
    };
    addToWatchList(data)
  }

  const removeFromWatchListHandler = () => {
    const data = {
      "media_type": "movie",
      "media_id": id,
      "watchlist": false
    };
    removeFromWatchList(data)
  }

  return (
    <div className="col-md-3 mt-2">
    <div className="card card-height p-0 col-12">
      <img src={image ? `https://image.tmdb.org/t/p/w500/${image}` : 'no-image.png'} className="card-img-top image-size mb-2" alt="..."/>
      <div className="d-flex flex-column justify-content-between card-body p-2">
        <h5 data-testid="movie-title" className="card-title">{truncate(title, 30)}</h5>
        <div>
          <p data-testid="movie-overview" className="card-text">{truncate(overview, 70)}</p>
        </div>
        <div className="align-self-center actions text-center">
          <Link data-testid="movie-details-link" to={`movie/${id}`}>See Details</Link>
          <div className="d-flex mt-2">
            {
              favouriteIDs.includes(id) ? <button data-testid="active-favourite" onClick={removeFromFavouritesHandler} className="btn mr-2"><i className="fa fa-star text-favourite"></i></button>
                : <button data-testid="inactive-favourite" onClick={makeFavouriteHandler} className="btn mr-2"><i className="fa fa-star"></i></button>
            }
            {
              watchlistIDs.includes(id) ? <button data-testid="active-watchlist" onClick={removeFromWatchListHandler} className="btn rm-btn-bg">Remove</button>
                : <button data-testid="inactive-watchlist" onClick={addToWatchListHandler} className="btn">Watch later</button>
            }
          </div>
        </div>
      </div>
    </div>
    </div>
  );
}

export default MovieCard;
