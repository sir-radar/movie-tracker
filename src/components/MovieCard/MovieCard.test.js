import React from 'react';
import { MemoryRouter } from "react-router-dom";
import { render, cleanup, fireEvent } from '@testing-library/react';
import MovieCard from './MovieCard';
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer"

afterEach(cleanup)

const renderComponent = (props) => {
  const div = document.createElement("div");
  return render(<MemoryRouter><MovieCard {...props}/></MemoryRouter>, div)
}

const propsData = {
  id: 1,
  title: "Movie title", 
  image: "", 
  overview: "Movie overview", 
  addToFavourite: jest.fn(), 
  removeFavourite: jest.fn(), 
  favouriteIDs: [1,2,3,4],
  removeFromWatchList: jest.fn(),
  addToWatchList: jest.fn(),
  watchlistIDs: [1,2,3,4]
};

describe('MovieCard Component', () => {

  it("renders without crashing", () => {
    renderComponent(propsData)
  })

  it("renders movie card details", () => {
    const { getByTestId } = renderComponent(propsData)
    expect(getByTestId('movie-title')).toHaveTextContent(propsData.title)
    expect(getByTestId('movie-overview')).toHaveTextContent(propsData.overview)
    expect(getByTestId('movie-details-link')).toHaveTextContent("See Details")
  })

  it("display favourite sign if movie is marked as favourite", () => {
    const { getByTestId, queryByTestId } = renderComponent(propsData)
    expect(getByTestId('active-favourite')).not.toBeNull()
    expect(queryByTestId('inactive-favourite')).toBeNull()
  })

  it("display unfavourite sign if movie is not marked as favourite", () => {
    const { queryByTestId } = renderComponent({...propsData, id:10})
    expect(queryByTestId('active-favourite')).toBeNull()
    expect(queryByTestId('inactive-favourite')).not.toBeNull()
  })

  it("should call removeFavourite after favourite sign is clicked", () => {
    const { queryByTestId } = renderComponent(propsData)
    fireEvent.click(queryByTestId('active-favourite'))
    expect(propsData.removeFavourite).toHaveBeenCalled()
  })

  it("should call addToFavourite after unfavourite sign is clicked", () => {
    const { queryByTestId } = renderComponent({...propsData, id:10})
    fireEvent.click(queryByTestId('inactive-favourite'))
    expect(propsData.addToFavourite).toHaveBeenCalled()
  })

  it("display active watchlist button if movie is marked as watchlist", () => {
    const { getByTestId, queryByTestId } = renderComponent(propsData)
    expect(getByTestId('active-watchlist')).not.toBeNull()
    expect(queryByTestId('inactive-watchlist')).toBeNull()
  })

  it("display inactive watchlist button if movie is not marked as watchlist", () => {
    const { queryByTestId } = renderComponent({...propsData, id:10})
    expect(queryByTestId('active-watchlist')).toBeNull()
    expect(queryByTestId('inactive-watchlist')).not.toBeNull()
  })

  it("should call removeFromWatchList after active watchlist button is clicked", () => {
    const { queryByTestId } = renderComponent(propsData)
    fireEvent.click(queryByTestId('active-watchlist'))
    expect(propsData.removeFromWatchList).toHaveBeenCalled()
  })

  it("should call addToWatchList after inactive watchlist button is clicked", () => {
    const { queryByTestId } = renderComponent({...propsData, id:10})
    fireEvent.click(queryByTestId('inactive-watchlist'))
    expect(propsData.addToWatchList).toHaveBeenCalled()
  })

  // it("navigates to movie deatils page when you click the movie details page link", () => {
  //   const { getByTestId } = renderComponent(propsData)
  //   fireEvent.click(getByTestId('movie-details-link'));
  //   expect(document.querySelector('[data-testid="movie-details-page"]')).not.toBeNull()
  // })

  it("matches snapshot", () => {
    const tree = renderer.create(<MemoryRouter><MovieCard {...propsData}/></MemoryRouter>).toJSON();
    expect(tree).toMatchSnapshot();
  })

});