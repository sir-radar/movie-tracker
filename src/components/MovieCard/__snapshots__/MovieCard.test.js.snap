// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="col-md-3 mt-2"
>
  <div
    className="card card-height p-0 col-12"
  >
    <img
      alt="..."
      className="card-img-top image-size mb-2"
      src="no-image.png"
    />
    <div
      className="d-flex flex-column justify-content-between card-body p-2"
    >
      <h5
        className="card-title"
        data-testid="movie-title"
      >
        Movie title
      </h5>
      <div>
        <p
          className="card-text"
          data-testid="movie-overview"
        >
          Movie overview
        </p>
      </div>
      <div
        className="align-self-center actions text-center"
      >
        <a
          data-testid="movie-details-link"
          href="/movie/1"
          onClick={[Function]}
        >
          See Details
        </a>
        <div
          className="d-flex mt-2"
        >
          <button
            className="btn mr-2"
            data-testid="active-favourite"
            onClick={[Function]}
          >
            <i
              className="fa fa-star text-favourite"
            />
          </button>
          <button
            className="btn rm-btn-bg"
            data-testid="active-watchlist"
            onClick={[Function]}
          >
            Remove
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
